Druntime fails on buildkite:
https://buildkite.com/dlang/dmd/builds/21165#d27bdd39-ff1a-4d7d-812e-b34a41e96c66/445-3393
```
generated/linux/release/64/benchmark.o:benchmark/runbench.d:_D32TypeInfo_S3std4file11DirIterator6__initZ: error: undefined reference to '_D3std4file11DirIterator9__xtoHashFNbNeKxSQBoQBnQBlZm'
generated/linux/release/64/benchmark.o:benchmark/runbench.d:_D32TypeInfo_S3std4file11DirIterator6__initZ: error: undefined reference to '_D3std4file11DirIterator11__xopEqualsFKxSQBnQBmQBkKxQmZb'
collect2: error: ld returned 1 exit status
```

Demangled symbols:
```
nothrow @trusted ulong std.file.DirIterator.__xtoHash(ref const(std.file.DirIterator))
bool std.file.DirIterator.__xopEquals(ref const(std.file.DirIterator), ref const(std.file.DirIterator))
```

I don't know why.

--- a/src/dmd/escape.d
+++ b/src/dmd/escape.d
@@ -1282,11 +1282,18 @@ private bool checkReturnEscapeImpl(Scope* sc, Expression e, bool refs, bool gag)
 
     foreach (VarDeclaration v; er.byref)
     {
-        if (log) printf("byref `%s`\n", v.toChars());
+        if (log)
+        {
+            printf("byref `%s`\n", v.toChars());
+            if (v.storage_class & STC.return_) printf(" return");
+            if (v.storage_class & STC.ref_)    printf(" ref");
+            if (v.storage_class & STC.scope_)  printf(" scope");
+            printf("\n");
+        }
 
         // 'featureState' tells us whether to emit an error or a deprecation,
         // depending on the flag passed to the CLI for DIP25
-        void escapingRef(VarDeclaration v, FeatureState featureState = FeatureState.enabled)
+        void escapingRef(VarDeclaration v, ScopeRef vsr, FeatureState featureState = FeatureState.enabled)
         {
             if (!gag)
             {
@@ -1295,7 +1302,9 @@ private bool checkReturnEscapeImpl(Scope* sc, Expression e, bool refs, bool gag)
                     (v.type.hasPointers() || v.storage_class & STC.ref_))
                 {
                     msg = "returning `%s` escapes a reference to parameter `%s`";
-                    supplemental = "perhaps annotate the parameter with `return`";
+                    supplemental = vsr == ScopeRef.Ref_ReturnScope
+                                              ? "perhaps remove `scope` parameter annotation so `return` applies to `ref`"
+                                              : "perhaps annotate the parameter with `return`";
                 }
                 else
                 {
@@ -1314,6 +1323,12 @@ private bool checkReturnEscapeImpl(Scope* sc, Expression e, bool refs, bool gag)
         if (v.isDataseg())
             continue;
 
+        auto stc = v.storage_class;
+        if (stc & STC.out_)
+            stc |= STC.ref_;  // temporary hack until we adjust buildScopeRef()
+
+        const vsr = buildScopeRef(refs, stc);
+
         Dsymbol p = v.toParent2();
 
         // https://issues.dlang.org/show_bug.cgi?id=19965
@@ -1324,7 +1339,7 @@ private bool checkReturnEscapeImpl(Scope* sc, Expression e, bool refs, bool gag)
         {
             if (p == sc.func)
             {
-                escapingRef(v);
+                escapingRef(v, vsr, FeatureState.enabled);
                 continue;
             }
             FuncDeclaration fd = p.isFuncDeclaration();
@@ -1342,15 +1357,18 @@ private bool checkReturnEscapeImpl(Scope* sc, Expression e, bool refs, bool gag)
                     sc.func.storage_class |= STC.return_ | STC.returninferred;
                 }
             }
-
         }
 
         /* Check for returning a ref variable by 'ref', but should be 'return ref'
          * Infer the addition of 'return', or set result to be the offending expression.
          */
-        if (v.isReference() && !(v.storage_class & (STC.return_ | STC.foreach_)))
+        if ((vsr == ScopeRef.Ref ||
+             vsr == ScopeRef.RefScope ||
+             vsr == ScopeRef.Ref_ReturnScope) &&
+            !(v.storage_class & STC.foreach_))
         {
-            if (sc.func.flags & FUNCFLAG.returnInprocess && p == sc.func)
+            if (sc.func.flags & FUNCFLAG.returnInprocess && p == sc.func &&
+                (vsr == ScopeRef.Ref || vsr == ScopeRef.RefScope))
             {
                 inferReturn(sc.func, v);        // infer addition of 'return'
             }
@@ -1362,7 +1380,7 @@ private bool checkReturnEscapeImpl(Scope* sc, Expression e, bool refs, bool gag)
                 {
                     //printf("escaping reference to local ref variable %s\n", v.toChars());
                     //printf("storage class = x%llx\n", v.storage_class);
-                    escapingRef(v, global.params.useDIP25);
+                    escapingRef(v, vsr, global.params.useDIP25);
                     continue;
                 }
                 // Don't need to be concerned if v's parent does not return a ref

--- a/src/dmd/mtype.d
+++ b/src/dmd/mtype.d
@@ -7233,6 +7233,7 @@ ScopeRef buildScopeRef(bool returnByRef, StorageClass stc) pure nothrow @nogc @s
     ScopeRef result;
     final switch (stc & (STC.ref_ | STC.scope_ | STC.return_))
     {
+        case STC.return_:
         case 0:                        result = ScopeRef.None;        break;
         case STC.ref_:                 result = ScopeRef.Ref;         break;
         case STC.scope_:               result = ScopeRef.Scope;       break;

--- a/test/fail_compilation/fail21868b.d
+++ b/test/fail_compilation/fail21868b.d
@@ -1,0 +1,22 @@
+/* REQUIRED_ARGS: -preview=dip1000
+TEST_OUTPUT:
+---
+fail_compilation/fail21868b.d(19): Error: returning `&s.x` escapes a reference to parameter `s`
+fail_compilation/fail21868b.d(19):        perhaps remove `scope` parameter annotation so `return` applies to `ref`
+---
+*/
+
+// https://issues.dlang.org/show_bug.cgi?id=21868
+
+struct S
+{
+    int x;
+    int* y;
+}
+
+int* test(ref return scope S s)
+{
+    return &s.x;
+}
+
+

